indexing-service/src/main/java/io/druid/indexing/common/task/TaskMonitor.java

```diff
@@ -0,0 +1,502 @@
+          numRunningTasks + numSucceededTasks + numFailedTasks,
+          expectedNumSucceededTasks
+      );
+    }
+  }
+
+  Set<String> getRunningTaskIds()
+  {
+    return runningTasks.values().stream().map(entry -> entry.runningTask.getId()).collect(Collectors.toSet());
+  }
+
+  List<SubTaskSpec<T>> getRunningSubTaskSpecs()
+  {
+    return runningTasks.values().stream().map(monitorEntry -> monitorEntry.spec).collect(Collectors.toList());
+  }
+
+  @Nullable
+  MonitorEntry getRunningTaskMonitorEntory(String subTaskSpecId)
+  {
+    return runningTasks.values()
+                       .stream()
+                       .filter(monitorEntry -> monitorEntry.spec.getId().equals(subTaskSpecId))
+                       .findFirst()
+                       .orElse(null);
+  }
+
+  List<SubTaskSpec<T>> getCompleteSubTaskSpecs()
+  {
+    return taskHistories.values().stream().map(history -> history.spec).collect(Collectors.toList());
+  }
+
+  @Nullable
+  TaskHistory<T> getCompleteSubTaskSpecHistory(String subTaskSpecId)
+  {
+    return taskHistories.get(subTaskSpecId);
+  }
+
+  class MonitorEntry
+  {
+    private final SubTaskSpec<T> spec;
+    private final T runningTask;
+    // old tasks to recent tasks. running task is not included
+    private final CopyOnWriteArrayList<TaskStatusPlus> taskHistory;
+    private final SettableFuture<SubTaskCompleteEvent<T>> completeEventFuture;
+
+    /**
+     * This variable is updated inside of the {@link java.util.concurrent.Callable} executed by
+     * {@link #taskStatusChecker}, and can be read by calling {@link #getRunningStatus}.
+     */
+    @Nullable
+    private volatile TaskStatusPlus runningStatus;
+
+    MonitorEntry(
+        SubTaskSpec<T> spec,
+        T runningTask,
+        @Nullable TaskStatusPlus runningStatus,
+        SettableFuture<SubTaskCompleteEvent<T>> completeEventFuture
+    )
+    {
+      this(spec, runningTask, runningStatus, new CopyOnWriteArrayList<>(), completeEventFuture);
+    }
+
+    private MonitorEntry(
+        SubTaskSpec<T> spec,
+        T runningTask,
+        @Nullable TaskStatusPlus runningStatus,
+        CopyOnWriteArrayList<TaskStatusPlus> taskHistory,
+        SettableFuture<SubTaskCompleteEvent<T>> completeEventFuture
+    )
+    {
+      this.spec = spec;
+      this.runningTask = runningTask;
+      this.runningStatus = runningStatus;
+      this.taskHistory = taskHistory;
+      this.completeEventFuture = completeEventFuture;
+    }
+
+    MonitorEntry withNewRunningTask(T newTask, @Nullable TaskStatusPlus newStatus, TaskStatusPlus statusOfLastTask)
+    {
+      taskHistory.add(statusOfLastTask);
+      return new MonitorEntry(
+          spec,
+          newTask,
+          newStatus,
+          taskHistory,
+          completeEventFuture
+      );
+    }
+
+    int numTries()
+    {
+      return taskHistory.size() + 1; // count runningTask as well
+    }
+
+    void updateStatus(TaskStatusPlus statusPlus)
```

Comment:
is there any memory barrier that would guarantee a write in one thread is picked up by a read in another?

Reply:
gotcha